File: frameworks/python/pandas/workloads.py
import pandas as pd

def project(df: pd.DataFrame):
    return df[["state", "votes"]]


def group_count(df: pd.DataFrame):
    return df.groupby("party").count()


def sort(df: pd.DataFrame):
    return df.sort_values("votes", ascending=False)


def flip_bool(df: pd.DataFrame):
    return df["won"].apply(lambda _bool: not _bool)


def find_max(df: pd.DataFrame):
    return df["votes"].max()


def select(df: pd.DataFrame):
    return df[(df["party"] == "DEM") & (df["won"]) & (df["votes"] > 20000)]


def select_group_mean(df: pd.DataFrame):
    return df[df["won"]].groupby("party").mean()


def longest_string_in_column(df: pd.DataFrame):
    return df["candidate"].apply(lambda name: len(name)).max()


def join(df: pd.DataFrame):
    return pd.merge(df, df, how="inner", on="unique_col")


def select_p90_cols(df: pd.DataFrame):
    return df.iloc[:, 1: round(df.shape[1] * 0.9)]


def select_p90_rows(df: pd.DataFrame):
    return df.iloc[1: round(df.shape[0] * 0.9)]


def combine_cols(df: pd.DataFrame):
    df["combined"] = df["candidate"] + " " + df["party"] + " " + df["won"].astype(str)
    return df

File: frameworks/r/datatable/workloads.R
library(data.table)

project <- function(df) {
  return(df[, c("state", "votes")])
}

group_count <- function(df){ 
  return(df[,.N, by=c("party")])
}

sort <- function(df){
  return(df[order(-votes)])
}

flip_bool <- function(df){
  return(df[, won := lapply(.SD, function(x) !x), .SDcols = "won"])
}

find_max <- function(df){
  return(df[,max(votes)])
}

select <- function(df){
  return(df[party == "DEM" & won & votes > 20000])
}

select_group_mean <- function(df){
  return(df[won==TRUE, mean(votes), by = "party"])
}

longest_string_in_column <- function(df){
  return(df[, max(nchar(candidate))])
}

join <- function(df){
  return(df[df, on="unique_col", nomatch=0])
}

select_p90_cols <- function(df){
  return(df[,1:round(ncol(df) * 0.9)])
}

select_p90_rows <- function(df){
  return(df[1:round(nrow(df) * 0.9)])
}

combine_cols <- function(df){
  return(df[ , combined := do.call(paste, c(.SD, sep = " ")), .SDcols = c("candidate", "party", "won")])
}
